chore: refactor `BuildProbeJoinMetrics` to use `BaselineMetrics`

Samyak2 opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #16495

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

Add a `BaselineMetrics` and remove `output_rows` from `BuildProbeJoinMetrics`.
- The `elapsed_compute` in baseline is populated in the Drop trait implementation by summing up `join_time` and `build_time`.

## Are these changes tested?

Here's an example of an `explain analyze` of a hash join showing these metrics:
```
[(WatchID@0, WatchID@0)], metrics=[output_rows=100, elapsed_compute=2.313624ms, build_input_batches=1, build_input_rows=100, input_batches=1, input_rows=100, output_batches=1, build_mem_used=3688, build_time=865.832µs, join_time=1.369875ms]
```

Notice `output_rows=100, elapsed_compute=2.313624ms` in the above.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

No

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

Samyak2 commented on datafusion/physical-plan/src/joins/cross_join.rs:
```diff
@@ -632,7 +632,7 @@ impl<T: BatchTransformer> CrossJoinStream<T> {
 
                     self.join_metrics.output_batches.add(1);
-                    self.join_metrics.output_rows.add(batch.num_rows());
+                    self.join_metrics.baseline.record_output(batch.num_rows());
```
I have updated it to use `record_poll`. Could you check if that's what you were looking for? It's my first time with this codebase, so I'm a little unsure if I did the right thing

Samyak2 commented:
Rebased on latest main

2010YOUY01 commented:
Thank you! this implementation looks correct to me.

Since the state transition in joins are tricky, could you add a test (or ensure there are some existing tests), to double-check this refactor won't change the result of the related metrics?

Samyak2 commented:
Actually, I see the same behavior on latest main. Looks like the output_rows metric in nested loop join is currently wrong?

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -825,7 +825,8 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+                    let poll = handle_state!(self.process_probe_batch());
+                    self.join_metrics.baseline.record_poll(poll)
                 }
                 NestedLoopJoinStreamState::ExhaustedProbeSide => {
```
I think this state is also possible to output final result: because if it's left join, NLJ has to exhaust right input first, then output the unmatched left rows.

2010YOUY01 commented:
> Actually, I see the same behavior on latest main. Looks like the output_rows metric in nested loop join is currently wrong?

Thank you for the catch! Here might also need `record_poll()`: https://github.com/apache/datafusion/pull/16500/files#r2187952197

Samyak2 commented:
> > Actually, I see the same behavior on latest main. Looks like the output_rows metric in nested loop join is currently wrong?
> 
> Thank you for the catch! Here might also need `record_poll()`: https://github.com/apache/datafusion/pull/16500/files#r2187952197

Thanks for the pointer! It makes sense now. I have added a record_poll there and also updated the `output_batches` metric inside the function. Please take a look

Samyak2 commented:
Fixed the formatting issues

2010YOUY01 reviewed (approved):
Thanks again.
